MoreLinq/Reactive/Subject.cs

```diff
@@ -0,0 +1,92 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 Atif Aziz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq.Reactive
+{
+    using System;
+    using System.Collections.Generic;
+    using Delegate = Delegating.Delegate;
+
+    sealed class Subject<T> : IObservable<T>, IObserver<T>
+    {
+        List<IObserver<T>> _observers;
+        bool _completed;
+        Exception _error;
+
+        bool HasObservers => (_observers?.Count ?? 0) > 0;
+        List<IObserver<T>> Observers => _observers ?? (_observers = new List<IObserver<T>>());
+
+        bool IsMuted => _completed || _error != null;
+
+        public IDisposable Subscribe(IObserver<T> observer)
+        {
+            if (observer == null) throw new ArgumentNullException(nameof(observer));
+
+            if (_error is Exception error)
+            {
+                observer.OnError(error);
+                return Disposable.Nop;
+            }
+
+            if (_completed)
+            {
+                observer.OnCompleted();
+                return Disposable.Nop;
+            }
+
+            Observers.Add(observer);
+            return Delegate.Disposable(() => Observers.Remove(observer));
+        }
+
+        public void OnNext(T value)
+        {
+            if (!HasObservers)
+                return;
+
+            foreach (var observer in Observers)
+                observer.OnNext(value);
+        }
+
+        public void OnError(Exception error)
+        {
+            if (IsMuted)
+                return;
+
+            _error = error;
+
+            if (!(Assignment.Set(ref _observers, default) is List<IObserver<T>> observers))
```

Comment:
> Why do we need to set _observers back to null?

Because no further events can occur after `OnError` or `OnCompleted`. Any future subscription after a subject has reached a concluded state can be communicated immediately on subscription without needing to keep the obvserver. Like Rx, protocol violations are not reported as exceptions.

> but maybe it would be nice to add a comment why?

Okay, done with 261e36037e49480c2b2f73d42721d1940b064f89.